lib/model/autocomplete.dart

```diff
@@ -1243,3 +1321,264 @@ class TopicAutocompleteResult extends AutocompleteResult {
+    // that ranking takes precedence over this.
+
+    final channelId = switch (narrow) {
+      ChannelNarrow(:var streamId) || TopicNarrow(:var streamId) => streamId,
+      DmNarrow() => null,
+      CombinedFeedNarrow()
+        || MentionsNarrow()
+        || StarredMessagesNarrow()
+        || KeywordSearchNarrow() => () {
+             assert(false, 'No compose box, thus no autocomplete is available in ${narrow.runtimeType}.');
+             return null;
+           }(),
+    };
+    return (a, b) => _compareByRelevance(a, b, currentChannelId: channelId);
+  }
+
+  // Check `typeahead_helper.compare_by_activity` in Zulip web;
+  // We follow the behavior of Web but with a small difference in that Web
+  // compares "recent activity" only for subscribed channels, but we do it
+  // for unsubscribed ones too.
+  //   https://github.com/zulip/zulip/blob/c3fdee6ed/web/src/typeahead_helper.ts#L972-L988
+  static int _compareByRelevance(ZulipStream a, ZulipStream b, {
+    required int? currentChannelId,
+  }) {
+    // The order of each comparator element in the list is important; the first
+    // one having the highest priority and the last one having the least priority.
+    final comparators = [
+      if (currentChannelId != null)
+        () => compareByBeingCurrent(a, b, currentChannelId: currentChannelId),
+      () => compareByBeingSubscribed(a, b),
+      () => compareByRecentActivity(a, b),
+      () => compareByWeeklyTraffic(a, b),
+      () => ChannelStore.compareChannelsByName(a, b),
+    ];
+    return comparators.map((compare) => compare())
+      .firstWhere((result) => result != 0, orElse: () => 0);
+  }
+
+  /// Comparator that puts the channel being composed to, before other ones.
+  @visibleForTesting
+  static int compareByBeingCurrent(ZulipStream a, ZulipStream b, {
+    required int currentChannelId,
+  }) {
+    return switch ((a.streamId, b.streamId)) {
+      (int id, _) when id == currentChannelId => -1,
+      (_, int id) when id == currentChannelId =>  1,
+      _                                       =>  0,
+    };
+  }
+
+  /// Comparator that puts subscribed channels before unsubscribed ones.
+  ///
+  /// For subscribed channels, it puts them in the following way:
+  ///   pinned unmuted > unpinned unmuted > pinned muted > unpinned muted
+  @visibleForTesting
+  static int compareByBeingSubscribed(ZulipStream a, ZulipStream b) {
+    return switch((tryCast<Subscription>(a), tryCast<Subscription>(b))) {
+      (Subscription(), null) => -1,
+      (null, Subscription()) =>  1,
+      (Subscription(isMuted:  false), Subscription(isMuted:   true)) => -1,
+      (Subscription(isMuted:   true), Subscription(isMuted:  false)) =>  1,
+      (Subscription(pinToTop:  true), Subscription(pinToTop: false)) => -1,
+      (Subscription(pinToTop: false), Subscription(pinToTop:  true)) =>  1,
+      _                                                              =>  0,
+    };
```

Comment:
I'd prefer not to add and use `tryCast` for this, and instead do something like:

```suggestion
    if (a is  Subscription && b is! Subscription) return -1;
    if (a is! Subscription && b is  Subscription) return 1;
    
    return switch((a, b)) {
      (Subscription(isMuted:  false), Subscription(isMuted:   true)) => -1,
      (Subscription(isMuted:   true), Subscription(isMuted:  false)) =>  1,
      (Subscription(pinToTop:  true), Subscription(pinToTop: false)) => -1,
      (Subscription(pinToTop: false), Subscription(pinToTop:  true)) =>  1,
      _                                                              =>  0,
    };
```

which is equivalent and doesn't add a step for the reader to interpret where `null` comes from and what it means. It also separates the main, headline logic from the rest, corresponding to the dartdoc's choice of what goes in its first line vs. the body:

```dart
  /// Comparator that puts subscribed channels before unsubscribed ones.
  ///
  /// For subscribed channels, it puts them in the following way:
  ///   pinned unmuted > unpinned unmuted > pinned muted > unpinned muted
```

Reply:
(also nit: s/way/order/ in that dartdoc)